graphql_compiler/compiler/emit_sql.py

```diff
@@ -0,0 +1,747 @@
+            name @output(out_name: "animal_or_descendant_name")
+        }
+    }
+}
+as an example, this is addressed with the following algorithm:
+
+1. Recursively collapse the query, treating the recursive component as a black box. For this
+example, this results in the rough SQL:
+
+SELECT
+    animal.name AS animal_name,
+    location.name AS location_name
+FROM
+    animal
+LEFT JOIN animal_livesin ON animal_livesin.animal_id = animal.animal_id
+LEFT JOIN location ON location.location_id = animal_livesin.livesin_id
+WHERE
+    animal.name IN :names
+
+2. Wrap this query as a CTE, and include any link columns in the output. A link column is the column
+that the recursive clause will later be attached to.
+
+WITH base_cte AS ( -- the actual name of the CTE is an anonymous table name
+    SELECT
+        animal.name as animal_name,
+        location.name as location_name
+        animal.animal_id as link_column -- the actual name of the column an anonymous column name
+    FROM
+        animal
+    LEFT JOIN animal_livesin ON animal_livesin.animal_id = animal.animal_id
+    LEFT JOIN location ON location.location_id = animal_livesin.livesin_id
+    WHERE
+        animal.name IN :names
+)
+
+3. Construct the recursive clause. Here we only recurse on the columns necessary to JOIN before
+and after the recursion, output columns are not carried along. The recursion is joined to the
+CTE of the base query, ensuring that the recursion only starts at the required starting points,
+no more.
+
+Also worth noting with the recursive clause is the __depth_internal_name, which keeps track of
+recursion depth per the compiler's semantics, and __path_internal_name, which is used for cycle
+detection. Simply using UNION here to join the anchor query to the recursive query is insufficient
+to prevent cycles, because the depth column is unique to each row, even in a cycle.
+
+
+WITH RECURSIVE recursive_cte AS (
+    -- anchor query
+    SELECT DISTINCT
+        animal_parentof.animal_id AS animal_id,
+        animal_parentof.parentof_id AS parentof_id,
+        0 AS __depth_internal_name,
+        CAST(animal_2.animal_id AS VARCHAR) || ',' AS __path_internal_name,
+        0 AS __cycle_detected_internal_name
+    FROM
+        animal_parentof
+        JOIN base_cte ON base_cte.link_column == animal_parentof.animal_id
+    UNION ALL
+    -- recursive query
+    SELECT
+        recursive_cte.animal_id,
+        animal_parentof.parentof_id,
+        -- increment the depth
+        recursive_cte.__depth_internal_name + 1 AS __depth_internal_name,
+        -- append the next element to the path
+        recursive_cte.__path_internal_name || CAST(animal_parentof.parentof_id AS VARCHAR) ||
+            ',' AS __path_internal_name,
+        -- check if a cycle is already present, or if this creates a cycle
+        CASE WHEN (recursive_cte.__cycle_detected_internal_name = 1) THEN 1
+             WHEN (recursive_cte.__path_internal_name LIKE '%' ||
```

Comment:
Consider parens around the right argument to `LIKE`? Otherwise, even if this is executed correctly by the SQL database, the order of operations and therefore the semantics are not obvious to the reader.